devito/mpi/utils.py

```diff
@@ -0,0 +1,147 @@
+        * ``fmapper``: ``function -> [(dimension, side, amount), ...]``
+
+    For the :class:`Dimension`s that need no halo exchange but appear in one
+    or more of the tracked :class:`TensorFunction`s, a further mapper
+
+        * ``fixed``: ``dimension -> dimension``
+
+    is provided telling at which index in these dimensions should the halo
+    be inserted.
+    For example, consider the :class:`Function` ``u(t, x, y)``. Assume ``x`` and
+    ``y`` require a halo exchange. A question is: once the halo exchange is
+    performed, at what offset in ``t`` should it be placed? should it be at
+    ``u(0, ...)`` or ``u(1, ...)`` or even ``u(t-1, ...)``? The ``fixed`` mapper
+    provides this information.
+    """
+
+    def __init__(self, exprs):
+        self._mapper = {}
+        self._fixed = {}
+
+        # What Functions actually need a halo exchange?
+        need_halo = as_mapper(Scope(exprs).d_all, lambda i: i.function)
+        need_halo = {k: v for k, v in need_halo.items() if k.is_TensorFunction}
+
+        for i in exprs:
+            for f, v in i.dspace.parts.items():
+                if f not in need_halo:
+                    continue
+                if f.grid is None:
+                    raise RuntimeError("`%s` needs a `Grid` for a HaloScheme" % f.name)
+                for d in f.dimensions:
+                    r = d.root
+                    if v[r].is_Null:
+                        continue
+                    elif d in f.grid.distributor.dimensions:
+                        # Found a distributed dimension, calculate what and how
+                        # much halo is needed
+                        lsize = f._offset_domain[d].left - v[r].lower
+                        if lsize > 0:
+                            self._mapper.setdefault(f, []).append((d, LEFT, lsize))
+                        rsize = v[r].upper - f._offset_domain[d].right
+                        if rsize > 0:
+                            self._mapper.setdefault(f, []).append((d, RIGHT, rsize))
+                    else:
+                        # Found a serial dimension, we need to determine where,
+                        # along this dimension, the halo will have to be placed
+                        fixed = self._fixed.setdefault(f, OrderedDict())
+                        shift = int(any(d in dep.cause for dep in need_halo[f]))
+                        # Examples:
+                        # u[t+1, x] = f(u[t, x])   => shift == 1
+                        # u[t-1, x] = f(u[t, x])   => shift == 1
+                        # u[t+1, x] = f(u[t+1, x]) => shift == 0
+                        # In the first and second cases, the x-halo should be inserted
+                        # at `t`, while in the last case it should be inserted at `t+1`
+                        if i.ispace.directions[r] is Forward:
+                            last = v[r].upper - shift
+                        else:
+                            last = v[r].lower + shift
```

Comment:
would put `+shift` everywhere with +-1 shifts personally

Reply:
this is changed anyway by my next PR

Reply:
EDIT: in fact, I completely rewrote this piece of code (for the sake of clarity), so even better